### PR TITLE
Revise Pickle Remote Task for Jupyter Notebook Environment

### PR DESCRIPTION
**Pull request by Mecoli1219**
## Tracking PR
https://github.com/flyteorg/flytekit/pull/2733

## Why are the changes needed?
As discussed in the comments of flyteorg/flytekit#2733, the current design of the Jupyter Notebook is not well considered.
image

## What changes were proposed in this pull request?
- [x] Add `basename` in `logger.debug`
- [x] Remove `__setstate__` and `__getstate__` as pickling `Promise` object is not necessary as expected
- [x] Redesign the interactive mode logic
  - [x] Pickling in the workflow level, so that we don't need extra kwargs in `Options`.
  - [x] Following with the previous point, the `interactive_mode_enabled` field in `SerailizationSettings` is not required either.
  - [x] Make SerializationSettings non-mutating by initializing the fast_registration_setting during construction in `remote.py`.
  - [x] Update `DefaultNotebookTaskResolver` to enable extracting the task entity from the pickled object. 
      - I updated the pickled object from entity to a dictionary of entity, where the key of the dictionary is the name of the task entity. 
  - [x] Automatic versioning for the interactive mode if the version is not provided in `remote.execute(...)` command.

### Setup process
Testing code is exactly the same as flyteorg/flytekit#2733

### Screenshots
| version 1 | version 2-1 | version 2-2 (run again without modification)|
| - | - | - |
| image | image | image |
| image | image | image |

Rerunning the code without modification and explicitly specify the `version` value will use the latest version in interactive mode. Note that `wf` workflow will register `hello` task, so the version of `hello` in second round is equal to the version of `wf` in the first round.

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
https://github.com/flyteorg/flytekit/pull/2733

### CODE REVIEW

File: flytekit/remote/remote.py [hunk 1]
```diff
@@ -2526,3 +2539,49 @@ def download(
...
+                        entity.set_resolver(default_notebook_task_resolver)
+                        pickled_target_dict[entity.name] = entity
+            elif isinstance(entity, WorkflowBase):
+                for task in entity.nodes:
+                    queue.append(task)
+            elif isinstance(entity, CoreNode):
+                queue.append(entity.flyte_entity)
+        return pickled_target_dict
+
+    def _pickle_entity(self, entity: typing.Any):
```
Comment (thomasjpfan) [inline flytekit/remote/remote.py:2568]:
```suggestion
    def _pickle_and_upload_entity(self, entity: typing.Any) -> FastSerializationSettings:
```

File: flytekit/core/python_auto_container.py [hunk 2]
```diff
@@ -292,16 +292,18 @@ def name(self) -> str:
...
     @timeit("Load task")
     def load_task(self, loader_args: List[str]) -> PythonAutoContainerTask:
+        _, entity_name, *_ = loader_args
         import gzip
 
         import cloudpickle
 
         with gzip.open(PICKLE_FILE_PATH, "r") as f:
-            return cloudpickle.load(f)
+            entity_dict = cloudpickle.load(f)
```
Comment (thomasjpfan) [inline flytekit/core/python_auto_container.py:301]: This means each task will load a bigger pickle object than it needs to and that each task will need more memory to load everything up.

But given how normal fast-register creates a tarball with all the task I am okay with this.

File: flytekit/remote/remote.py (hunk 1 above)
Comment (Mecoli1219) [reply flytekit/remote/remote.py]: updated

**Comment by Mecoli1219** [comment]
@wild-endeavor I updated the automatic versioning followed with the same logistic as [FlyteRemote.register_script](https://github.com/flyteorg/flytekit/blob/51751ebb5b2aaa62264e094b8987d86214350283/flytekit/remote/remote.py#L1124C9-L1145C14). If we are in interactive mode, we will automatically compute a hashed `version` for it if version is not explicitly specified. Otherwise, we will follow the original behavior in `FlyteRemote`.

**Review by wild-endeavor — APPROVED** [review]
thanks!!

**Pull request merged**